File: ToyLanguage7/ToyLanguage7/FileTable.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System;
using System.IO;


namespace ToyLanguage7
{
    class FileTable<K, V> : MyIFileDictionary<K, V>
    {
        private Dictionary<K, V> dict;

        public FileTable() { this.dict = new Dictionary<K, V>(); }

        public void add(K id, V val)
        {
            dict.Add(id, val);
        }

        public V get(K key)
        {
            return dict[key];
        }

        public bool isDefined(K id)
        {
            return dict.ContainsKey(id);
        }

        public bool isEmpty()
        {
            return dict.Count() == 0;
        }

        public V lookup(K id)
        {
            if (isDefined(id))
                return get(id);
            else
            {
                throw new MyException("variable " + id + " is not defined");
            }
        }

        public void remove(K key)
        {
            dict.Remove(key);
        }

        public int size()
        {
            return dict.Count();
        }

        public void update(K id, V val)
        {
            dict[id] = val;
        }
        
        public override string ToString()
        {
            string result = "";
            Dictionary<K, V>.KeyCollection keyColl = dict.Keys;
            foreach (K k in keyColl)
            {
                result = result + k + ":";
                //if (get(k) is KeyValuePair<string, StreamReader>)
                //    result += get(k).ToString();
                result += get(k);
                result += Environment.NewLine;
            }
            return result;
        }
    }
}

File: ToyLanguage4/src/Model/FileTable.java
package Model;

import javafx.util.Pair;

import java.io.BufferedReader;
import java.util.HashMap;

public class FileTable implements MyIFileDictionary<Integer,Pair<String, BufferedReader>>
{
    private int key=0;
    private HashMap<Integer,Pair<String,BufferedReader>> dict;


    public FileTable() {
        this.dict = new HashMap<>();
    }

    @Override
    public boolean isDefined(Integer id) {
        return dict.containsKey(id);
    }

    @Override
    public void update(Integer id, Pair<String, BufferedReader> val) {
        key++;
        dict.put(id,val);
        //return key++;
    }

    @Override
    public void add(Pair<String, BufferedReader> val) {
        key++;
        dict.put(key,val);
        //return key++;
    }

    @Override
    public Pair<String, BufferedReader> lookup(Integer id) throws MyException {
        if(isDefined(id))
            return get(id);
        else
        {
            throw new MyException("variable " + id +" is not defined!");
        }
    }

    @Override
    public int size() {
        return dict.size();
    }

    @Override
    public void remove(Integer key) {
        dict.remove(key);
        this.key--;
    }

    @Override
    public boolean isEmpty() {
        return dict.size()==0;
    }

    @Override
    public Pair<String, BufferedReader> get(Integer key) {
        return dict.get(key);
    }

    public int getKey() {
        return key;
    }

    @Override
    public String toString() {
        StringBuilder result;
        result = new StringBuilder();
        for(int d: this.dict.keySet())
        {
            result.append(d).append(":");
            result.append(this.dict.get(d).getKey());
            result.append("\n");
        }

        if(result.length()==0)
            return "";
        return result.substring(0,result.length()-1);
        //return key+"";
    }
}
